src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseIdentityReport.java

```diff
@@ -0,0 +1,233 @@
+    }
+
+    public String getKey() {
+      return key;
+    }
+
+    /**
+     * Assumes that PRECISE samples will always have _nn_ in the name, and that Tissue Origin will
+     * never be "nn". Should hopefully be a safe assumption since this is a biobanking project where
+     * tracking tissue origin is critical.
+     */
+    public Predicate<SampleDto> predicate() {
+      return sample -> {
+        if (SAMPLE_CATEGORY_IDENTITY.equals(sample.getSampleType())) return false;
+        return timePointCode == getTimesReceived(sample.getName());
+      };
+    }
+  }
+
+  @Override
+  protected void collectData(PineryClient pinery) throws HttpResponseException, IOException {
+    List<SampleDto> allPreciseSamples =
+        pinery
+            .getSample()
+            .allFiltered(new SampleClient.SamplesFilter().withProjects(Lists.newArrayList("PRE")));
+    List<String> tissueOriginOrder = Lists.newArrayList("_Se_", "_Pl_", "_Ly_");
+    List<String> tissueOriginOrderExtra =
+        Lists.newArrayList("_Se_", "_Pl_", "_Ly_", "_Us_", "_Up_");
+    allPreciseIdentities =
+        allPreciseSamples
+            .stream()
+            .filter(s -> s.getSampleType().equals(SAMPLE_CATEGORY_IDENTITY))
+            .collect(Collectors.toList());
+
+    table = new LinkedList<>();
+
+    // Add headers
+    table.add(
+        Lists.newArrayList(
+            "ID",
+            "Screening/Randomization Serum",
+            "Screening/Randomization Plasma",
+            "Screening/Randomization Buffy Coat",
+            "Screening/Randomization Urine Supernatant",
+            "Screening/Randomization Urine Pellet",
+            "6 mo Follow Up Serum",
+            "6 mo Follow Up Plasma",
+            "6 mo Follow Up Buffy Coat",
+            "1 yr Follow Up Serum",
+            "1 yr Follow Up Plasma",
+            "1 yr Follow Up Buffy Coat",
+            "1 yr Follow Up Urine Supernatant",
+            "1 yr Follow Up Urine Pellet",
+            "18 mo Follow Up Serum",
+            "18 mo Follow Up Plasma",
+            "18 mo Follow Up Buffy Coat",
+            "2 yr Follow Up Serum",
+            "2 yr Follow Up Plasma",
+            "2 yr Follow Up Buffy Coat",
+            "2 yr Follow Up Urine Supernatant",
+            "2 yr Follow Up Urine Pellet",
+            "Bx Slides - Positive",
+            "Bx Slides - Negative",
+            "RP Slides"));
+
+    // One row per identity
+    for (SampleDto identity : allPreciseIdentities) {
+      List<String> row = new LinkedList<>();
+      row.add(identity.getName());
+
+      // For all regular time points
+      for (int i = 0; i < 5; i++) {
+        TimePoint currentTimePoint = TimePoint.values()[i];
+        List<SampleDto> samplesAtCurrentTime =
+            allPreciseSamples
+                .stream()
+                .filter(
+                    s ->
+                        s.getName().startsWith(identity.getName())) // Get children of this Identity
```

Comment:
Never parse the name for information if it can be avoided. You can use `SampleUtils#getParent(String, String, Map<String, SampleDto)` to find the sample's Identity. `SampleUtils#mapSamplesById` is also useful to create the necessary map.

Reply:
Hi Dillan, in this line I have an identity (that's s) and need to get all of its children. It doesn't look like there's a 'getChildren' to mirror getParent, is there another way to achieve this or should I implement a 'getChildren'?

Reply:
There is no `getChildren` method. In the report I'm working on, I'm getting all of the descendants for each identity using this method:

```
  private Map<String, List<SampleDto>> mapChildrenByIdentityId() {
    Map<String, List<SampleDto>> childrenByIdentityId = new HashMap<>();
    for (SampleDto sample : samplesById.values()) {
      if (!SAMPLE_CATEGORY_IDENTITY.equals(getAttribute(ATTR_CATEGORY, sample))) {
        SampleDto identity = getParent(sample, SAMPLE_CATEGORY_IDENTITY, samplesById);
        if (!childrenByIdentityId.containsKey(identity.getId())) {
          childrenByIdentityId.put(identity.getId(), new ArrayList<>());
        }
        childrenByIdentityId.get(identity.getId()).add(sample);
      }
    }
    return childrenByIdentityId;
  }
```

(and then I have another list filtered to include only the identities)

You could do something similar, or go the other way, using `SampleDto#getChildren` and working your way down the tree. Going up is usually simpler though due to branches and variable hierarchy

Reply:
Thanks! Updated